File: examples/dump_sumcheck.rs
// ===========================================================================
// dump_sumcheck.rs — Sumcheck 协议 golden test。
//
// 运行: cargo run --example dump_sumcheck --release > golden_sumcheck_rs.txt
// 对拍: diff golden_sumcheck_rs.txt golden_sumcheck_cpp.txt
//
// 手动构造 sponge 输入 (protocol_id/session_id/c0/c2), 绕过 WHIR DomainSeparator
// 的内部 CBOR map 序列化差异, 确保与 C++ 端字节级对拍。
// ===========================================================================

use ark_ff::{AdditiveGroup, PrimeField, Zero};
use sha3::{Digest, Sha3_256, Sha3_512};
use spongefish::DuplexSpongeInterface;
use whir::algebra::fields::Field64;
use whir::algebra::sumcheck::{compute_sumcheck_polynomial, fold};
use whir::transcript::codecs::U64;

fn dump_bytes(label: &str, data: &[u8]) {
    print!("  {label} ");
    for byte in data { print!("{byte:02x}"); }
    println!();
}

struct Lcg(u64);
impl Lcg {
    fn new(seed: u64) -> Self { Self(seed) }
    fn next_u64(&mut self) -> u64 {
        self.0 = self.0.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        self.0
    }
    fn next_field64(&mut self) -> Field64 { Field64::from(self.next_u64()) }
}

fn main() {
    let mut rng = Lcg::new(0xAAAABBBBCCCCDDDDu64);

    let initial_size = 8usize;
    let num_rounds = 3usize;

    let a: Vec<Field64> = (0..initial_size).map(|_| rng.next_field64()).collect();
    let b: Vec<Field64> = (0..initial_size).map(|_| rng.next_field64()).collect();

    println!("# SECTION sumcheck");
    print!("  vector_a");
    for v in &a { print!(" {}", v); }
    println!();
    print!("  vector_b");
    for v in &b { print!(" {}", v); }
    println!();

    let mut sum = Field64::zero();
    for i in 0..initial_size { sum += a[i] * b[i]; }
    println!("  initial_sum {}", sum);

    // ---- 构造 sponge 输入 (与 C++ 端完全一致的显式字节) ----
    // protocol_id = SHA3-512(cbor(0xABCD)) = SHA3-512([19 AB CD])
    let mut h = Sha3_512::new();
    ciborium::into_writer(&0xABCDu32, &mut h).unwrap();
    let pid: [u8; 64] = h.finalize().into();

    // session_id = SHA3-256(cbor("sumcheck_dump")) = SHA3-256([6D + 14 bytes])
    let mut h = Sha3_256::new();
    ciborium::into_writer(&"sumcheck_dump".to_string(), &mut h).unwrap();
    let sid: [u8; 32] = h.finalize().into();

    // ---- Prover ----
    let mut a_prove = a.clone();
    let mut b_prove = b.clone();
    let mut sum_prove = sum;

    // 手动模拟 WHIR ProverState: sponge + narg
    let mut sponge = spongefish::instantiations::Shake128::default();
    let mut narg_buf: Vec<u8> = vec![];

    // absorb DomainSeparator
    sponge.absorb(&pid);
    sponge.absorb(&sid);

    println!("CASE 0 sumcheck prove");

    for round in 0..num_rounds {
        let (c0, c2) = compute_sumcheck_polynomial(&a_prove, &b_prove);
        println!("  round{round}_c0 {}", c0);
        println!("  round{round}_c2 {}", c2);

        // prover_message(c0): encode → absorb → narg
        let c0_bytes = c0.into_bigint().as_ref()[0].to_le_bytes();
        sponge.absorb(&c0_bytes);
        use spongefish::NargSerialize;
        U64(c0.into_bigint().as_ref()[0]).serialize_into_narg(&mut narg_buf);

        // prover_message(c2): encode → absorb → narg
        let c2_bytes = c2.into_bigint().as_ref()[0].to_le_bytes();
        sponge.absorb(&c2_bytes);
        U64(c2.into_bigint().as_ref()[0]).serialize_into_narg(&mut narg_buf);

        // verifier_message<Field64>
        let mut r_bytes = [0u8; 8];
        sponge.squeeze(&mut r_bytes);
        let r = Field64::from(u64::from_le_bytes(r_bytes));
        println!("  round{round}_r {}", r);

        fold(&mut a_prove, r);
        fold(&mut b_prove, r);
        let c1 = sum_prove - c0.double() - c2;
        sum_prove = (c2 * r + c1) * r + c0;
        println!("  round{round}_sum {}", sum_prove);
    }

    println!("CASE 1 final vectors");
    print!("  final_a");
    for v in &a_prove { print!(" {}", v); }
    println!();
    print!("  final_b");
    for v in &b_prove { print!(" {}", v); }
    println!();

    dump_bytes("proof_narg", &narg_buf);

    // ---- Verifier ----
    let mut vs_sponge = spongefish::instantiations::Shake128::default();
    vs_sponge.absorb(&pid);
    vs_sponge.absorb(&sid);

    use spongefish::NargDeserialize;
    let mut narg_cursor: &[u8] = &narg_buf;
    let mut sum_verify = sum;

    println!("CASE 2 sumcheck verify");

    for round in 0..num_rounds {
        // prover_message: deserialize → absorb
        let c0u = U64::deserialize_from_narg(&mut narg_cursor).unwrap();
        let c0 = Field64::from(c0u.0);
        let c2u = U64::deserialize_from_narg(&mut narg_cursor).unwrap();
        let c2 = Field64::from(c2u.0);

        vs_sponge.absorb(&c0u.0.to_le_bytes());
        vs_sponge.absorb(&c2u.0.to_le_bytes());

        println!("  round{round}_c0 {}", c0);
        println!("  round{round}_c2 {}", c2);

        let mut r_bytes = [0u8; 8];
        vs_sponge.squeeze(&mut r_bytes);
        let r = Field64::from(u64::from_le_bytes(r_bytes));
        println!("  round{round}_r {}", r);

        let c1 = sum_verify - c0.double() - c2;
        sum_verify = (c2 * r + c1) * r + c0;
        println!("  round{round}_sum {}", sum_verify);
    }

    let eof = narg_cursor.is_empty();
    println!("  check_eof {}", eof as i32);
}

File: cpp/tests/dump_sumcheck.cpp
// ===========================================================================
// dump_sumcheck.cpp — Sumcheck 协议 golden test。
//
// 运行: ./dump_sumcheck > golden_sumcheck_cpp.txt
// 对拍: diff golden_sumcheck_rs.txt golden_sumcheck_cpp.txt
//
// 测试: sumcheck prove → verify 完整周期, 与 Rust 端字节级对拍。
// ===========================================================================

#include "whir/algebra/goldilocks.hpp"
#include "whir/algebra/sumcheck.hpp"
#include "whir/protocols/proof_of_work.hpp"
#include "whir/protocols/sumcheck_protocol.hpp"
#include "whir/transcript/transcript.hpp"

#include <cstdint>
#include <cstdio>
#include <vector>

using F = ::whir::algebra::Goldilocks;

// LCG — 与 Rust 侧完全一致
struct Lcg { uint64_t s; explicit Lcg(uint64_t seed) : s(seed) {}
    uint64_t next() { s = s * 6364136223846793005ULL + 1442695040888963407ULL; return s; }
};

// 打印 Goldilocks 值为十进制 u64 (匹配 Rust Display)
void print_f(const char* label, F v) {
    std::printf("  %s %llu\n", label, (unsigned long long)v.as_canonical_u64());
}

// 十六进制打印字节数组
template <typename T>
void dump_bytes(const char* label, const T& data) {
    std::printf("  %s ", label);
    for (auto byte : data) std::printf("%02x", static_cast<unsigned>(byte));
    std::printf("\n");
}

int main() {
    Lcg rng(0xAAAABBBBCCCCDDDDULL);

    const std::size_t initial_size = 8;  // 2^3 = 8
    const std::size_t num_rounds = 3;

    // 确定性向量
    std::vector<F> a(initial_size);
    std::vector<F> b(initial_size);
    for (std::size_t i = 0; i < initial_size; ++i) a[i] = F::from_u64(rng.next());
    for (std::size_t i = 0; i < initial_size; ++i) b[i] = F::from_u64(rng.next());

    std::printf("# SECTION sumcheck\n");

    std::printf("  vector_a");
    for (auto& v : a) std::printf(" %llu", (unsigned long long)v.as_canonical_u64());
    std::printf("\n");
    std::printf("  vector_b");
    for (auto& v : b) std::printf(" %llu", (unsigned long long)v.as_canonical_u64());
    std::printf("\n");

    // 内积 sum_i a[i]*b[i]
    F sum = F::zero();
    for (std::size_t i = 0; i < initial_size; ++i) sum += a[i] * b[i];
    std::printf("  initial_sum %llu\n", (unsigned long long)sum.as_canonical_u64());

    // Config
    ::whir::protocols::sumcheck::Config<F> config;
    config.initial_size = initial_size;
    config.num_rounds = num_rounds;
    config.round_pow = ::whir::protocols::pow::PowConfig{};

    // 手动计算 protocol_id / session_id (与 Rust 一致)
    ::whir::transcript::DomainSeparator ds;
    {
        uint8_t cbor_proto[] = {0x19, 0xAB, 0xCD};
        sha3_512_hash(cbor_proto, 3, ds.protocol_id.data());
        uint8_t cbor_sess[] = {0x6D, 0x73, 0x75, 0x6D, 0x63, 0x68, 0x65, 0x63, 0x6B, 0x5F, 0x64, 0x75, 0x6D, 0x70};
        sha3_256_hash(cbor_sess, 14, ds.session_id.data());
    }

    // ---- Prover ----
    auto a_prove = a;
    auto b_prove = b;
    F sum_prove = sum;

    ::whir::transcript::Empty instance;
    auto ps = ::whir::transcript::ProverState::from_ds(ds, instance);

    std::printf("CASE 0 sumcheck prove\n");

    for (std::size_t round = 0; round < num_rounds; ++round) {
        // 计算 sumcheck 多项式系数 c0, c2
        auto [c0, c2] = ::whir::algebra::compute_sumcheck_polynomial<F>(a_prove, b_prove);
        print_f((std::string("round") + std::to_string(round) + "_c0").c_str(), c0);
        print_f((std::string("round") + std::to_string(round) + "_c2").c_str(), c2);

        ps.prover_message(c0);
        ps.prover_message(c2);

        F r = ps.template verifier_message<F>();
        print_f((std::string("round") + std::to_string(round) + "_r").c_str(), r);

        // 折叠向量: a, b 各压缩一半
        ::whir::algebra::fold<F>(a_prove, r);
        ::whir::algebra::fold<F>(b_prove, r);
        // c1 = sum - 2*c0 - c2, 新 sum = c(r)
        F c1 = sum_prove - (c0 + c0) - c2;
        sum_prove = (c2 * r + c1) * r + c0;
        print_f((std::string("round") + std::to_string(round) + "_sum").c_str(), sum_prove);
    }

    // 最终向量
    std::printf("CASE 1 final vectors\n");
    std::printf("  final_a");
    for (auto& v : a_prove) std::printf(" %llu", (unsigned long long)v.as_canonical_u64());
    std::printf("\n");
    std::printf("  final_b");
    for (auto& v : b_prove) std::printf(" %llu", (unsigned long long)v.as_canonical_u64());
    std::printf("\n");

    auto proof = std::move(ps).proof();
    dump_bytes("proof_narg", proof.narg_string);

    // ---- Verifier ----
    auto vs = ::whir::transcript::VerifierState::from_ds(ds, instance, proof);
    F sum_verify = sum;

    std::printf("CASE 2 sumcheck verify\n");

    for (std::size_t round = 0; round < num_rounds; ++round) {
        F c0, c2;
        vs.prover_message(c0);
        vs.prover_message(c2);
        print_f((std::string("round") + std::to_string(round) + "_c0").c_str(), c0);
        print_f((std::string("round") + std::to_string(round) + "_c2").c_str(), c2);

        F r = vs.template verifier_message<F>();
        print_f((std::string("round") + std::to_string(round) + "_r").c_str(), r);

        F c1 = sum_verify - (c0 + c0) - c2;
        sum_verify = (c2 * r + c1) * r + c0;
        print_f((std::string("round") + std::to_string(round) + "_sum").c_str(), sum_verify);
    }

    bool eof = vs.check_eof();
    std::printf("  check_eof %d\n", static_cast<int>(eof));

    return 0;
}
